src/EnergyPlus/WaterThermalTanks.cc

```diff
@@ -8112,8 +8124,95 @@ namespace WaterThermalTanks {
+                    }
+                }
+
+                // Make adjustments to A and B to account for heaters being on or off now
+                if (Tank.HeaterOn1 and !PrevHeaterOn1) {
+                    // If heater 1 is on now and wasn't before add the heat rate to the B term
+                    B[Tank.HeaterNode1 - 1] += Qheater1 / (Tank.Node(Tank.HeaterNode1).Mass * Cp);
+                } else if (!Tank.HeaterOn1 and PrevHeaterOn1) {
+                    // If heater 1 is off now and was on before, remove the heat rate from the B term
+                    B[Tank.HeaterNode1 - 1] -= Tank.MaxCapacity / (Tank.Node(Tank.HeaterNode1).Mass * Cp);
+                }
+                if (Tank.HeaterOn2 and !PrevHeaterOn2) {
+                    // If heater 2 is on now and wasn't before add the heat rate to the B term
+                    B[Tank.HeaterNode2 - 1] += Qheater2 / (Tank.Node(Tank.HeaterNode2).Mass * Cp);
+                } else if (!Tank.HeaterOn2 and PrevHeaterOn2) {
+                    // If heater 1 is off now and was on before, remove the heat rate from the B term
+                    B[Tank.HeaterNode2 - 1] -= Tank.MaxCapacity / (Tank.Node(Tank.HeaterNode2).Mass * Cp);
+                }
+
+                if ((Tank.HeaterOn1 || Tank.HeaterOn2) and !(PrevHeaterOn1 || PrevHeaterOn2)) {
+                    // Remove off cycle loads
+                    // Apply on cycle loads
+                    for (int i = 0; i < nTankNodes; i++) {
+                        auto &node(Tank.Node[i]);
+                        Real64 NodeCapacitance = node.Mass * Cp;
+                        A[i] += (node.OffCycLossCoeff - node.OnCycLossCoeff) / NodeCapacitance;
+                        B[i] += (- node.OffCycParaLoad + node.OnCycParaLoad + (node.OnCycLossCoeff - node.OffCycLossCoeff) * Tank.AmbientTemp) / NodeCapacitance;
+                    }
+                } else if (!(Tank.HeaterOn1 || Tank.HeaterOn2) and (PrevHeaterOn1 || PrevHeaterOn2)) {
+                    // Remove on cycle loads
+                    // Apply off cycle loads
+                    for (int i = 0; i < nTankNodes; i++) {
+                        auto &node(Tank.Node[i]);
+                        Real64 NodeCapacitance = node.Mass * Cp;
+                        A[i] -= (node.OffCycLossCoeff - node.OnCycLossCoeff) / NodeCapacitance;
+                        B[i] -= (- node.OffCycParaLoad + node.OnCycParaLoad + (node.OnCycLossCoeff - node.OffCycLossCoeff) * Tank.AmbientTemp) / NodeCapacitance;
+                    }
+                }
+
+                // Set the sub timestep (dt)
+                dt = TimeRemaining;
+                for (int i = 0; i < nTankNodes; ++i) {
+                    const Real64 Denominator = fabs(A[i] * Tavg[i] + B[i]);
+                    if (Denominator != 0.0)
+                        dt = min(dt, dT_max / Denominator);
+                }
+                dt = max(min(SubTimestepMin, TimeRemaining), dt);
```

Comment:
How often will TimeRemaining be less than SubTimestepMin (10 sec)? How much impact would it have to allow TimeRemaining to go down to 1 sec as the previous model allowed?

Reply:
I have that backwards. TimeRemaining is less than SubTimestepMin when these iterations are near completion. It's hard to understand this at first glance. Nearing the end of the iterations, TimeRemaining is approaching 0 and dT_max is also approaching 0. So the max(min()) above simply caps dt at 10 sec or TimeRemaining, whichever is less and takes the greater of that or dt. Then the line below limits dt to SubTimeStepMax. So when is dt a small number?

Reply:
`dT_max` may or may not approach zero by the end of `TimeRemaining`. It's a measure of how much we'll allow the tank temperature to change before doing the control decisions and looping around again. We want to loop and recalculate about the time a heater kicks on or off. Every once in a while that happens near the end of a system time step. This logic makes sure we don't go past the end of the system timestep but right up to it. This line is confusing. Maybe I should rewrite it to be more clear. 

I messed around with `dt` being allowed to go down to 1 second, but it counteracted our performance gains. The estimation of the next `dt` tends to undershoot it because it is a linear approximation of a non-linear solution. This means that without a lower bound on `dt` it would never get there. A 1 second lower bound works, but it takes a lot of decreasingly small steps before being pushed over. Each time around the sub timestep loop is more expensive now as there are more computations than with the old model, so it canceled it out. 

I'd argue that the 10 second minimum sub timestep is at least as good as a solution as the previous 1 second because it uses the analytical solution over that longer time rather than the linear forward Euler approximation the old model used. The only place where we'd see a difference is in making a control decision for a heater. It may happen a few seconds later with this model. 

That being said this is an improvement over what we have in 9.1, which is that the sub timestep is _always_ 60 seconds.